pkg/unshare/unshare.c

```diff
@@ -151,30 +154,89 @@ static char **parse_proc_stringlist(const char *list) {
 	return ret;
 }
 
-static int containers_reexec(void) {
-	char **argv, *exename;
+/*
+ * Taken from the runc cloned_binary.c file
+ * Copyright (C) 2019 Aleksa Sarai <cyphar@cyphar.com>
+ * Copyright (C) 2019 SUSE LLC
+ *
+ * This work is dual licensed under the following licenses. You may use,
+ * redistribute, and/or modify the work under the conditions of either (or
+ * both) licenses.
+ *
+ * === Apache-2.0 ===
+ */
+static int try_bindfd(void)
+{
+	int fd, ret = -1;
+	char src[PATH_MAX] = {0};
+	char template[64] = {0};
+
+	strncpy(template, "/tmp/containers.XXXXXX", sizeof(template) - 1);
+
+	/*
+	 * We need somewhere to mount it, mounting anything over /proc/self is a
+	 * BAD idea on the host -- even if we do it temporarily.
+	 */
+	fd = mkstemp(template);
+	if (fd < 0)
+		return ret;
+	close(fd);
+
+	ret = -EPERM;
+
+	if (readlink("/proc/self/exe", src, sizeof (src) - 1) < 0)
+		goto out;
+
+	if (mount(src, template, NULL, MS_BIND, NULL) < 0)
+		goto out;
+	if (mount(NULL, template, NULL, MS_REMOUNT | MS_BIND | MS_RDONLY, NULL) < 0)
+		goto out_umount;
+
+	/* Get read-only handle that we're sure can't be made read-write. */
+	ret = open(template, O_PATH | O_CLOEXEC);
+
+out_umount:
+	/*
+	 * Make sure the MNT_DETACH works, otherwise we could get remounted
+	 * read-write and that would be quite bad (the fd would be made read-write
+	 * too, invalidating the protection).
+	 */
+	if (umount2(template, MNT_DETACH) < 0) {
+		if (ret >= 0)
+			close(ret);
+		ret = -ENOTRECOVERABLE;
+	}
+
+out:
+	/*
+	 * We don't care about unlink errors, the worst that happens is that
+	 * there's an empty file left around in STATEDIR.
+	 */
+	unlink(template);
+	return ret;
+}
+
+static int copy_self_proc_exe(char **argv) {
+	char *exename;
 	int fd, mmfd, n_read, n_written;
 	struct stat st;
 	char buf[2048];
 
-	argv = parse_proc_stringlist("/proc/self/cmdline");
-	if (argv == NULL) {
-		return -1;
-	}
 	fd = open("/proc/self/exe", O_RDONLY | O_CLOEXEC);
 	if (fd == -1) {
 		fprintf(stderr, "open(\"/proc/self/exe\"): %m\n");
 		return -1;
 	}
 	if (fstat(fd, &st) == -1) {
 		fprintf(stderr, "fstat(\"/proc/self/exe\"): %m\n");
+		close(fd);
 		return -1;
 	}
 	exename = basename(argv[0]);
 	mmfd = syscall(SYS_memfd_create, exename, (long) MFD_ALLOW_SEALING | MFD_CLOEXEC);
 	if (mmfd == -1) {
 		fprintf(stderr, "memfd_create(): %m\n");
-		return -1;
+		goto close_fd;
```

Comment:
Why don't you use the autoclose fd stuff you use in crun?

Reply:
I've only adapted the code to the existing pattern coming from `try_bindfd`.

In any case, that is a good idea, I can work on cleaning up the C code and use auto cleanup functions.  Podman can also benefit from it.